File: engine/examples/resources/generate_test_font.py
#!/usr/bin/env python3
"""
Generate a simple 8x8 test font with perfect grid alignment.
Creates a 128x128 PNG (16x16 grid of 8x8 glyphs).
"""

from PIL import Image, ImageDraw

# Create 128x128 image (16x16 grid of 8x8 glyphs)
# This gives us 256 possible glyphs (full extended ASCII)
img_size = 128
chars_per_row = 16
char_size = 8

img = Image.new('RGB', (img_size, img_size), color=(0, 0, 0))
draw = ImageDraw.Draw(img)

# Define some simple test patterns for key characters
def draw_glyph(draw, x, y, char_code):
    """Draw a simple glyph at grid position (x, y)"""
    base_x = x * char_size
    base_y = y * char_size

    # Define patterns for specific characters
    if char_code == ord(' '):  # Space
        pass  # Leave blank

    elif char_code == ord('H'):  # H
        # Two vertical lines with horizontal crossbar
        for dy in range(8):
            draw.point((base_x + 1, base_y + dy), fill=(255, 255, 255))
            draw.point((base_x + 6, base_y + dy), fill=(255, 255, 255))
        for dx in range(2, 6):
            draw.point((base_x + dx, base_y + 4), fill=(255, 255, 255))

    elif char_code == ord('E'):  # E
        for dy in range(8):
            draw.point((base_x + 1, base_y + dy), fill=(255, 255, 255))
        for dx in range(1, 7):
            draw.point((base_x + dx, base_y + 0), fill=(255, 255, 255))
            draw.point((base_x + dx, base_y + 4), fill=(255, 255, 255))
            draw.point((base_x + dx, base_y + 7), fill=(255, 255, 255))

    elif char_code == ord('L'):  # L
        for dy in range(8):
            draw.point((base_x + 1, base_y + dy), fill=(255, 255, 255))
        for dx in range(1, 7):
            draw.point((base_x + dx, base_y + 7), fill=(255, 255, 255))

    elif char_code == ord('O'):  # O
        # Draw a rectangle
        for dy in range(1, 7):
            draw.point((base_x + 1, base_y + dy), fill=(255, 255, 255))
            draw.point((base_x + 6, base_y + dy), fill=(255, 255, 255))
        for dx in range(2, 6):
            draw.point((base_x + dx, base_y + 1), fill=(255, 255, 255))
            draw.point((base_x + dx, base_y + 6), fill=(255, 255, 255))

    elif char_code == ord('W'):  # W
        for dy in range(7):
            draw.point((base_x + 1, base_y + dy), fill=(255, 255, 255))
            draw.point((base_x + 6, base_y + dy), fill=(255, 255, 255))
        for dy in range(4, 8):
            draw.point((base_x + 3, base_y + dy), fill=(255, 255, 255))
            draw.point((base_x + 4, base_y + dy), fill=(255, 255, 255))

    elif char_code == ord('R'):  # R
        # Left vertical line
        for dy in range(8):
            draw.point((base_x + 1, base_y + dy), fill=(255, 255, 255))
        # Top horizontal
        for dx in range(1, 6):
            draw.point((base_x + dx, base_y + 0), fill=(255, 255, 255))
        # Middle horizontal
        for dx in range(1, 6):
            draw.point((base_x + dx, base_y + 4), fill=(255, 255, 255))
        # Right side top
        for dy in range(1, 4):
            draw.point((base_x + 6, base_y + dy), fill=(255, 255, 255))
        # Diagonal leg
        for i in range(3):
            draw.point((base_x + 4 + i, base_y + 5 + i), fill=(255, 255, 255))

    elif char_code == ord('D'):  # D
        # Left vertical line
        for dy in range(8):
            draw.point((base_x + 1, base_y + dy), fill=(255, 255, 255))
        # Top and bottom
        for dx in range(1, 6):
            draw.point((base_x + dx, base_y + 0), fill=(255, 255, 255))
            draw.point((base_x + dx, base_y + 7), fill=(255, 255, 255))
        # Right side
        for dy in range(1, 7):
            draw.point((base_x + 6, base_y + dy), fill=(255, 255, 255))

    elif char_code == ord('0'):  # 0
        # Similar to O
        for dy in range(1, 7):
            draw.point((base_x + 1, base_y + dy), fill=(255, 255, 255))
            draw.point((base_x + 6, base_y + dy), fill=(255, 255, 255))
        for dx in range(2, 6):
            draw.point((base_x + dx, base_y + 1), fill=(255, 255, 255))
            draw.point((base_x + dx, base_y + 6), fill=(255, 255, 255))

    elif char_code >= ord('1') and char_code <= ord('9'):  # Numbers 1-9
        # Simple vertical line for all numbers (simplified)
        for dy in range(8):
            draw.point((base_x + 3, base_y + dy), fill=(255, 255, 255))
            draw.point((base_x + 4, base_y + dy), fill=(255, 255, 255))

    elif char_code == ord('!'):  # !
        for dy in range(5):
            draw.point((base_x + 3, base_y + dy), fill=(255, 255, 255))
            draw.point((base_x + 4, base_y + dy), fill=(255, 255, 255))
        draw.point((base_x + 3, base_y + 7), fill=(255, 255, 255))
        draw.point((base_x + 4, base_y + 7), fill=(255, 255, 255))

    elif char_code == ord('.'):  # .
        draw.point((base_x + 3, base_y + 7), fill=(255, 255, 255))
        draw.point((base_x + 4, base_y + 7), fill=(255, 255, 255))

    else:
        # For all other characters, draw a simple test pattern
        # A small square in the center
        for dy in range(3, 6):
            for dx in range(3, 6):
                draw.point((base_x + dx, base_y + dy), fill=(255, 255, 255))

# Generate glyphs for ASCII 0-255
for i in range(256):
    grid_x = i % chars_per_row
    grid_y = i // chars_per_row
    draw_glyph(draw, grid_x, grid_y, i)

# Save the image
output_path = 'test_font_8x8.png'
img.save(output_path)
print(f"Generated {output_path}: {img_size}x{img_size} pixels, {chars_per_row}x{chars_per_row} grid of {char_size}x{char_size} glyphs")
print(f"Total glyphs: {chars_per_row * chars_per_row}")

File: engine/examples/generate_test_font.cpp
#include <SDL.h>
#include <SDL_image.h>
#include <iostream>
#include <vector>

// Generate a simple 8x8 test font with perfect grid alignment
// Creates a 128x128 PNG (16x16 grid of 8x8 glyphs)

void drawGlyph(SDL_Surface* surface, int gridX, int gridY, int charCode) {
    const int charSize = 8;
    int baseX = gridX * charSize;
    int baseY = gridY * charSize;

    Uint32* pixels = static_cast<Uint32*>(surface->pixels);
    Uint32 white = SDL_MapRGB(surface->format, 255, 255, 255);

    auto setPixel = [&](int dx, int dy) {
        if (dx >= 0 && dx < charSize && dy >= 0 && dy < charSize) {
            pixels[(baseY + dy) * surface->w + (baseX + dx)] = white;
        }
    };

    if (charCode == ' ') {
        // Space - leave blank
    }
    else if (charCode == 'H') {
        // Two vertical lines with horizontal crossbar
        for (int dy = 0; dy < 8; dy++) {
            setPixel(1, dy);
            setPixel(6, dy);
        }
        for (int dx = 2; dx < 6; dx++) {
            setPixel(dx, 4);
        }
    }
    else if (charCode == 'E') {
        for (int dy = 0; dy < 8; dy++) {
            setPixel(1, dy);
        }
        for (int dx = 1; dx < 7; dx++) {
            setPixel(dx, 0);
            setPixel(dx, 4);
            setPixel(dx, 7);
        }
    }
    else if (charCode == 'L') {
        for (int dy = 0; dy < 8; dy++) {
            setPixel(1, dy);
        }
        for (int dx = 1; dx < 7; dx++) {
            setPixel(dx, 7);
        }
    }
    else if (charCode == 'O') {
        // Draw a rectangle
        for (int dy = 1; dy < 7; dy++) {
            setPixel(1, dy);
            setPixel(6, dy);
        }
        for (int dx = 2; dx < 6; dx++) {
            setPixel(dx, 1);
            setPixel(dx, 6);
        }
    }
    else if (charCode == 'W') {
        for (int dy = 0; dy < 7; dy++) {
            setPixel(1, dy);
            setPixel(6, dy);
        }
        for (int dy = 4; dy < 8; dy++) {
            setPixel(3, dy);
            setPixel(4, dy);
        }
    }
    else if (charCode == 'R') {
        // Left vertical line
        for (int dy = 0; dy < 8; dy++) {
            setPixel(1, dy);
        }
        // Top horizontal
        for (int dx = 1; dx < 6; dx++) {
            setPixel(dx, 0);
        }
        // Middle horizontal
        for (int dx = 1; dx < 6; dx++) {
            setPixel(dx, 4);
        }
        // Right side top
        for (int dy = 1; dy < 4; dy++) {
            setPixel(6, dy);
        }
        // Diagonal leg
        for (int i = 0; i < 3; i++) {
            setPixel(4 + i, 5 + i);
        }
    }
    else if (charCode == 'D') {
        // Left vertical line
        for (int dy = 0; dy < 8; dy++) {
            setPixel(1, dy);
        }
        // Top and bottom
        for (int dx = 1; dx < 6; dx++) {
            setPixel(dx, 0);
            setPixel(dx, 7);
        }
        // Right side
        for (int dy = 1; dy < 7; dy++) {
            setPixel(6, dy);
        }
    }
    else if (charCode == '0') {
        // Similar to O
        for (int dy = 1; dy < 7; dy++) {
            setPixel(1, dy);
            setPixel(6, dy);
        }
        for (int dx = 2; dx < 6; dx++) {
            setPixel(dx, 1);
            setPixel(dx, 6);
        }
    }
    else if (charCode >= '1' && charCode <= '9') {
        // Simple vertical line for all numbers (simplified)
        for (int dy = 0; dy < 8; dy++) {
            setPixel(3, dy);
            setPixel(4, dy);
        }
    }
    else if (charCode == '!') {
        for (int dy = 0; dy < 5; dy++) {
            setPixel(3, dy);
            setPixel(4, dy);
        }
        setPixel(3, 7);
        setPixel(4, 7);
    }
    else if (charCode == '.') {
        setPixel(3, 7);
        setPixel(4, 7);
    }
    else {
        // For all other characters, draw a simple test pattern
        // A small square in the center
        for (int dy = 3; dy < 6; dy++) {
            for (int dx = 3; dx < 6; dx++) {
                setPixel(dx, dy);
            }
        }
    }
}

int main(int argc, char* argv[]) {
    if (SDL_Init(SDL_INIT_VIDEO) < 0) {
        std::cerr << "SDL_Init failed: " << SDL_GetError() << std::endl;
        return 1;
    }

    if (!(IMG_Init(IMG_INIT_PNG) & IMG_INIT_PNG)) {
        std::cerr << "IMG_Init failed: " << IMG_GetError() << std::endl;
        SDL_Quit();
        return 1;
    }

    // Create 128x128 image (16x16 grid of 8x8 glyphs)
    const int imgSize = 128;
    const int charsPerRow = 16;
    const int charSize = 8;

    SDL_Surface* surface = SDL_CreateRGBSurface(0, imgSize, imgSize, 32,
                                                 0xFF000000, 0x00FF0000, 0x0000FF00, 0x000000FF);

    if (!surface) {
        std::cerr << "Failed to create surface: " << SDL_GetError() << std::endl;
        IMG_Quit();
        SDL_Quit();
        return 1;
    }

    // Fill with black
    SDL_FillRect(surface, nullptr, SDL_MapRGB(surface->format, 0, 0, 0));

    SDL_LockSurface(surface);

    // Generate glyphs for ASCII 0-255
    for (int i = 0; i < 256; i++) {
        int gridX = i % charsPerRow;
        int gridY = i / charsPerRow;
        drawGlyph(surface, gridX, gridY, i);
    }

    SDL_UnlockSurface(surface);

    // Save the image
    const char* outputPath = "../examples/resources/test_font_8x8.png";
    if (IMG_SavePNG(surface, outputPath) != 0) {
        std::cerr << "Failed to save PNG: " << IMG_GetError() << std::endl;
        SDL_FreeSurface(surface);
        IMG_Quit();
        SDL_Quit();
        return 1;
    }

    std::cout << "Generated " << outputPath << ": " << imgSize << "x" << imgSize
              << " pixels, " << charsPerRow << "x" << charsPerRow
              << " grid of " << charSize << "x" << charSize << " glyphs" << std::endl;
    std::cout << "Total glyphs: " << (charsPerRow * charsPerRow) << std::endl;

    SDL_FreeSurface(surface);
    IMG_Quit();
    SDL_Quit();

    return 0;
}
